### PR TITLE
[WIP] Update ServiceBindingRequest.Status.Conditions when service not found

### PR DESCRIPTION
**Pull request by DhritiShikhar**
**This PR is on hold. Based on discussion with Predrag, I will pick up this PR in next sprint.**

### Motivation

When `Service` is not available, update `status.conditions` as
```
  conditions:
  - lastHeartbeatTime: "2020-05-28T20:07:51Z"
    lastTransitionTime: "2020-05-28T20:07:26Z"
    message: service not found
    reason: ServiceNotFound
    status: "False"
    type: Ready
```

### Changes

[1] Adds a check for Service Not Found case.

### How to test

1. Follow the steps given in this [example](https://github.com/redhat-developer/service-binding-operator/tree/master/examples/nodejs_postgresql) to:
- install DB operator, Service Binding Operator
- Import an application
- Create a DB instance 

2. Create the following `ServiceBindingRequest`:

```
apiVersion: apps.openshift.io/v1alpha1
kind: ServiceBindingRequest
metadata:
  name: binding-request
  namespace: service-binding-demo
spec:
  applicationSelector:
        resourceRef: nodejs-rest-http-crud 
        group: apps
        version: v1
        resource: deployments
 backingServiceSelector:
    group: postgresql.baiju.dev
    version: v1alpha1
    kind: Database
    resourceRef: db-demo
```

3. Use the following command to check if the status has been updated:
`oc get ServiceBindingRequest binding-request -o yaml`

The output should have `status.conditions.type` as 
-  `BindingReady` with `status: "False"`

```
status:
  conditions:
  - lastHeartbeatTime: "2020-05-28T20:07:51Z"
    lastTransitionTime: "2020-05-28T20:07:26Z"
    message: service not found
    reason: ServiceNotFound
    status: "False"
    type: Ready
  secret: ""

```

### Note

This PR does not deal with the scenario where `Service` is deleted and service is re-created with the same name.

### CODE REVIEW

**Comment by Avni-Sharma** [comment]
According to this https://github.com/redhat-developer/service-binding-operator/pull/498 we won't be having BindingReady. In that case we can have changes related to conditions when service not found  in #498 itself IMO.

**Comment by DhritiShikhar** [comment]
@Avni-Sharma I will rebase according to whichever PR gets in first.

**Comment by Avni-Sharma** [comment]
> @Avni-Sharma I will rebase according to whichever PR gets in first.

It means that either way both PRs are interrelated(since it is also depending as to which goes in first), so in that case, it would be good if it goes in as a single PR

**Comment by DhritiShikhar** [comment]
> It means that either way both PRs are interrelated(since it is also depending as to which goes in first), 

If it's about rebasing then in that case, every PR in this project is interrelated.
Are you suggesting that rebasing a PR is not appropriate?

> so in that case, it would be good if it goes in as a single PR

The only thing similar between this PR and #498 is that they both concern around `Status`.

This PR handles the case when a `Service` is not found.
#498 's concern is around `Application`.

I think the two PRs are modular enough and can go in separately.

File: pkg/controller/servicebindingrequest/reconciler.go [hunk 1]
```diff
@@ -155,6 +156,19 @@ func (r *Reconciler) Reconcile(request reconcile.Request) (reconcile.Result, err
 		r.RestMapper,
 	)
 	if err != nil {
+		if err == ErrServiceNotFound {
+			conditionsv1.SetStatusCondition(&sbr.Status.Conditions, conditionsv1.Condition{
+				Type:    BindingReady,
+				Status:  corev1.ConditionFalse,
+				Reason:  ServiceNotFoundReason,
+				Message: ErrServiceNotFound.Error(),
```
[the remark this answers is not in the document]
Comment (DhritiShikhar) [reply pkg/controller/servicebindingrequest/reconciler.go:164]: Yes we could. Let me check more on this and get back to you.

**Comment by DhritiShikhar** [comment]
@pedjak 

This PR should not be on hold for #512 
The correct PR is #498

**Comment by pedjak** [comment]
sorry @DhritiShikhar , my bad

/hold cancel

File: pkg/controller/servicebindingrequest/reconciler.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (DhritiShikhar) [reply pkg/controller/servicebindingrequest/reconciler.go]: not every backing service has a reference number

**Comment by Avni-Sharma** [comment]
@DhritiShikhar  what happens when we have multiple services and one of them is present and the other one isn't?

File: pkg/controller/servicebindingrequest/reconciler.go [hunk 2]
```diff
@@ -157,6 +158,19 @@ func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, err
 		r.restMapper,
 	)
 	if err != nil {
+		if err == errServiceNotFound {
+			conditionsv1.SetStatusCondition(&sbr.Status.Conditions, conditionsv1.Condition{
+				Type:    BindingReady,
+				Status:  corev1.ConditionFalse,
+				Reason:  ServiceNotFoundReason,
```
Comment (Avni-Sharma) [inline pkg/controller/servicebindingrequest/reconciler.go:165]: We can have a slice of services for multiple backing services, wherein we know which service is found and which service is not found, and then flag the readiness as False

Comment (DhritiShikhar) [reply pkg/controller/servicebindingrequest/reconciler.go:165]: Kind of like we have for applications?

```
status:
  applications:
  - group: apps
    kind: Deployment
    name: nodejs-rest-http-crud
    version: v1
```

File: pkg/controller/servicebindingrequest/reconciler_test.go [hunk 3]
```diff
@@ -400,6 +400,45 @@ func TestEmptyApplicationSelector(t *testing.T) {
 	require.Equal(t, 0, len(sbrOutput.Status.Applications))
 }
 
+// TestEmptyApplicationSelector tests that Status is successfully updated when Service is not found
+func TestServiceNotFound(t *testing.T) {
```
Comment (isutton) [inline pkg/controller/servicebindingrequest/reconciler_test.go:404]: Please adjust this to a test table with the following cases:

* Reconcile without a service, to validate whether the `BindingReady` condition is `False`; and
* Reconcile with a service, to validate whether `BindingReady` condition is `True`.

Please observe any other conditions that might be checked as well.

**Comment by isutton** [comment]
> @DhritiShikhar  what happens when we have multiple services and one of them is present and the other one isn't? 

It depends whether we are taking a "best effort" or an "all or nothing" approach.

The best effort approach could be collect as many services as possible, and populate the SBR secret with the collected values; the way the user is informed by this could be yet another condition (IncompleteServices) with a proper reason (ServiceNotFoundReason) and a message containing the name (or names) of the missing services.

The all or nothing approach could be stop at the first missing service, state this in the `CollectionReady` condition with the proper reason (ServiceNotFoundReason) and not create the SBR's secret.

I believe the latter might be better suited for now.

**Comment by pedjak** [comment]
@DhritiShikhar can we close this PR now, given the work in https://github.com/redhat-developer/service-binding-operator/pull/681 ?

**Comment by DhritiShikhar** [comment]
Yes.

**Pull request closed without merging**